ipa-core/src/cli/crypto/hybrid_encrypt.rs

```diff
@@ -71,28 +89,199 @@ impl HybridEncryptArgs {
             panic!("could not load network file")
         };
 
-        let shares: [Vec<HybridReport<BreakdownKey, TriggerValue>>; 3] =
-            input.iter::<TestHybridRecord>().share();
+        let mut worker_pool = ReportWriter::new(key_registries, &self.output_dir);
+        for (report_id, record) in input.iter::<TestHybridRecord>().enumerate() {
+            worker_pool.submit(report_id, record.share())?;
+        }
+
+        worker_pool.join()?;
+
+        let elapsed = start.elapsed();
+        tracing::info!(
+            "Encryption process is completed. {}s",
+            elapsed.as_secs_f64()
+        );
+
+        Ok(())
+    }
+}
 
-        for (index, (shares, key_registry)) in zip(shares, key_registries).enumerate() {
-            let output_filename = format!("helper{}.enc", index + 1);
-            let mut writer = OpenOptions::new()
+/// A thread-per-core pool responsible for encrypting reports in parallel.
+/// This pool is shared across all writers to reduce the number of context switches.
+struct EncryptorPool {
+    pool: Vec<(Sender<EncryptorInput>, JoinHandle<UnitResult>)>,
+    next_worker: usize,
+}
+
+impl EncryptorPool {
+    pub fn with_worker_threads(
+        thread_count: usize,
+        file_writer: [Sender<EncryptorOutput>; 3],
+        key_registries: [KeyRegistry<PublicKeyOnly>; 3],
+    ) -> Self {
+        Self {
+            pool: (0..thread_count)
+                .map(move |i| {
+                    let (tx, rx) = channel::<EncryptorInput>();
+                    let key_registries = key_registries.clone();
+                    let file_writer = file_writer.clone();
+                    (
+                        tx,
+                        std::thread::Builder::new()
+                            .name(format!("encryptor-{i}"))
+                            .spawn(move || {
+                                for (i, helper_id, report) in rx {
+                                    let key_registry = &key_registries[helper_id];
+                                    let output = report.encrypt(
+                                        DEFAULT_KEY_ID,
+                                        key_registry,
+                                        &mut thread_rng(),
+                                    )?;
+                                    file_writer[helper_id].send((i, output))?;
+                                }
+
+                                Ok(())
+                            })
+                            .unwrap(),
+                    )
+                })
+                .collect(),
+            next_worker: 0,
```

Comment:
why is `next_worker` always 0? what's the point of it if it's constant?

Reply:
mutation occurs inside `encrypt_share`